elf_load.c

```diff
@@ -311,13 +328,228 @@ static int elf_load_imgcpy(struct elf_prog *elf_prog, Elf *elf,
+	 * generating a page fault, because
+	 * vaend - vastart < PAGE_SIZE
+	 * ...
+	 */
+	memset((void *)vastart, 0, PAGE_ALIGN_UP(vastart) - vastart);
+
+	if (vaend == PAGE_ALIGN_UP(vastart))
+		return 0;
+
+	/*
+	 * ...
+	 * 2. vaend - vastart >= PAGE_SIZE, which can happen if
+	 * this segment contains a NOBITS section, such as .bss.
+	 * If that is the case, then simply anonymously map this
+	 * remaining area so that we don't waste time memsetting
+	 * it (.bss is quite large).
+	 */
+	vastart = PAGE_ALIGN_UP(vastart);
+	vastart = (uintptr_t)mmap((void *)vastart, vaend - vastart,
+				  PROT_EXEC | PROT_READ | PROT_WRITE,
+				  MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS,
+				  -1, 0);
+	if (unlikely(vastart == (uintptr_t)MAP_FAILED)) {
+		uk_pr_err("Failed to mmap the NOBITS part of phdr at "
+			  "offset %lu\n", phdr->p_offset);
+		return (int)vastart;
+	}
+
+	return 0;
+}
+
+/* Use this to mmap first PT_LOAD */
+static int do_elf_load_fdphdr_0(struct elf_prog *elf_prog,
+				GElf_Phdr *phdr, int fd)
+{
+	uintptr_t vastart_old, vastart_aligned;
+	uintptr_t vaend_old, vaend_new;
+	uintptr_t vastart, vaend;
+	__sz mmap_len;
+	int rc;
+
+	/* First start/vabase can't be !0 before loading first PT_LOAD */
+	UK_ASSERT(!elf_prog->start && !elf_prog->vabase && phdr);
+	/* PT_LOAD p_vaddr/p_paddr must be multiple of page size */
+	UK_ASSERT(!(phdr->p_vaddr & (phdr->p_align - 1)));
+	UK_ASSERT(!(phdr->p_paddr & (phdr->p_align - 1)));
+
+	mmap_len = prog_upperl + prog_lowerl + pt_load_max_align;
+
+	vastart = (uintptr_t)mmap(NULL, mmap_len,
+				  PROT_EXEC | PROT_READ | PROT_WRITE,
+				  MAP_PRIVATE,
+				  fd, phdr->p_offset);
+	if (unlikely(vastart == (uintptr_t)MAP_FAILED)) {
+		uk_pr_err("Failed to mmap the phdr at offset %lu\n",
+			  phdr->p_offset);
+		return (int)vastart;
+	}
+
+	vastart_aligned = ALIGN_UP(vastart, pt_load_max_align);
+	vastart_old = vastart;
+
+	/* Force remap with MAP_FIXED */
+	if (vastart_aligned != vastart) {
+		vastart = (uintptr_t)mmap((void *)vastart_aligned,
+					  phdr->p_filesz,
+					  PROT_EXEC | PROT_READ | PROT_WRITE,
+					  MAP_PRIVATE | MAP_FIXED,
+					  fd, phdr->p_offset);
+		if (unlikely(vastart == (uintptr_t)MAP_FAILED)) {
+			uk_pr_err("Failed to mmap the phdr at offset %lu\n",
+				  phdr->p_offset);
+			return (int)vastart;
+		}
+
+		rc = munmap((void *)vastart_old, vastart - vastart_old);
+		if (unlikely(rc)) {
+			uk_pr_err("Failed to unmap [vastart_old, vastart)\n");
+			return rc;
+		}
+	}
+
+	uk_pr_debug("%s: Memory mapped 0x%"PRIx64" - 0x%"PRIx64" to 0x%"PRIx64" - 0x%"PRIx64"\n",
+		    elf_prog->name,
+		    (uint64_t)phdr->p_offset,
+		    (uint64_t)phdr->p_offset + phdr->p_filesz,
+		    (uint64_t)vastart,
+		    (uint64_t)vastart + (uint64_t)phdr->p_filesz);
+
+	vaend_old = PAGE_ALIGN_UP(vastart_old + mmap_len);
+	vaend_new = PAGE_ALIGN_UP(vastart + phdr->p_filesz);
+	if (vaend_old - vaend_new) {
+		rc = munmap((void *)vaend_new, vaend_old - vaend_new);
```

Comment:
Isn't this what happens though? Also, I wish I did not have to do this because the following `mmap`s would use `MAP_FIXED` which should overwrite anyway, but as I said in my commit message:
> NOTE: The first segment must be mmap'd with the length of the entire
file in memory and then the surplus is to be `munmapped`. This way we
ensure we will always have a contiguous mapping big enough ready.
Although future `mmap`'s with `MAP_FIXED` should normally overwrite
this surplus, ridding us of the need to explicitly `munmap` it, in
this current state of Unikraft (0.14.1) there is a bug in `ukvmem`
that may not handle overlapping VMA's correctly.

Reply:
Yes, that's correct. Sorry for missing that.

Reply:
Hmm, definetely. However not even `ukvmem` itself is synchronised yet :laughing:. But I definetely see it in the future :disguised_face:.